src/store/Store.js

```diff
@@ -42,22 +43,50 @@ class Store {
       throw new Error('patchStrategy must be one of the included patching strategies or a custom patching function');
     }
 
+    this.maxReconnects = maxReconnects;
     this.portName = portName;
     this.readyResolved = false;
-    this.readyPromise = new Promise(resolve => this.readyResolve = resolve);
+    this.readyPromise = new Promise(resolve => {
+      this.readyResolve = resolve;
+    });
 
     this.browserAPI = getBrowserAPI();
     this.extensionId = extensionId; // keep the extensionId as an instance variable
-    this.port = this.browserAPI.runtime.connect(this.extensionId, {name: portName});
+    this.serializedMessageSender = withSerializer(serializer)((...args) => {
+      this.browserAPI.runtime.sendMessage(...args);
+    }, 1);
+    this.listeners = [];
+    this.state = state;
+    this.patchStrategy = patchStrategy;
+
     this.safetyHandler = this.safetyHandler.bind(this);
     if (this.browserAPI.runtime.onMessage) {
       this.safetyMessage = this.browserAPI.runtime.onMessage.addListener(this.safetyHandler);
     }
-    this.serializedPortListener = withDeserializer(deserializer)((...args) => this.port.onMessage.addListener(...args));
-    this.serializedMessageSender = withSerializer(serializer)((...args) => this.browserAPI.runtime.sendMessage(...args), 1);
-    this.listeners = [];
-    this.state = state;
-    this.patchStrategy = patchStrategy;
+    this.dispatch = this.dispatch.bind(this); // add this context to dispatch
+
+    // finally connect
+    this.connect(deserializer);
+  }
+
+  /**
+   * Connects and creates a port for handling messages between background and
+   * this store. Automatically reconnects when port becomes disconnected.
+   */
+  connect(deserializer, attempts = 0) {
+    if (attempts > this.maxReconnects) {
+      throw new Error("Too many connection attempts");
+    }
+    // wake up service worker and then connect
+    this.port = this.browserAPI.runtime.connect(this.extensionId, { name: this.portName });
+
+    this.port.onDisconnect.addListener(() => {
+      setTimeout(() => this.connect(deserializer, ++attempts), 0);
```

Comment:
When I attempted this approach, the major problem was that the background script stopped after processing the connection request.

Therefore, it causes the following problems:
- The store fails to sync after the reconnects is reached
- The resource consumption of the extension significantly increases as the reconnect applies to each tab. 

May I ask how you managed to solve these issues that I couldn't overcome with this architecture?

Reply:
> When I attempted this approach, the major problem was that the background script stopped after processing the connection request.

Yes, you cannot expect the service worker to be alive, they are transient. I have experienced a bug where two service workers may be started and the extension fails to sync (typically when the devtools arent' open). We have had no issues re-connecting in other cases though and the extension always restarts the service worker without fail.

> The store fails to sync after the reconnects is reached

I have not seen this, although our extension right now is not running through content scripts. Only the popup.html, but if i have multiple popup.html open, stopping the service worker manually is handled gracefully on all tabs and they begin re-syncing state.

> The resource consumption of the extension significantly increases as the reconnect applies to each tab.

I don't see how this increases resource consumption significantly other than the CPU to restart the connection. Set the maxReconnects to something smaller and handle in the script when re-connect fails.

> May I ask how you managed to solve these issues that I couldn't overcome with this architecture?

Anything that needs to survive past service worker restarts must be persisted to disk either e.g. chrome.storage.local or for sensitive data use [chrome.storage.session](https://developer.chrome.com/docs/extensions/reference/storage/#property-session)

Hope that helps!

Reply:
I feel this approach and the nature of the service worker will conflict with our extension for the downside highlighted in my previous post so we went for a different architecture as described on https://github.com/tshaddix/webext-redux/pull/282.

However, I am glad that you manage to sort your problems out through a different route. Please keep us informed if your user experiences any problems